bot/cogs/cyber.py

```diff
@@ -63,6 +63,10 @@ def __init__(self, bot: Bot):
             r"^.*\bwhat\b.*\belite\b.*\bemail\b.*$",
             re.IGNORECASE
         )
+        self.game_level_regex = re.compile(
+            r"^(le?v?e?l?).*?(\d{1,2}).*?(\d{1,2})\b.*$",
```

Comment:
Okay, so this regular expression has changed a lot. It matches this, which I think it shouldn't:

> **L**ol, remember the **2018** intake?

`:level 20 18`

This would maybe be acceptable, but in addition, the command will report out-of-range errors back to the user, in the form of a spammy message. We need to suppress these error messages if the command was invoked via this pattern match rather than a direct invocation of the command.

![iThe bot replies with "Invalid challenge number!" if asked for level 20 19](https://user-images.githubusercontent.com/18113170/52674965-1edf0480-2f1d-11e9-8025-9bb400b9f885.png)

<hr>

I think, also, in https://github.com/CyberDiscovery/cyberdisc-bot/pull/135/commits/54517984e7b282bfe49ecf55ff9338dafdb8ede5 we accidentally removed the `.*` prefixing the pattern for "level". This is bad because you can no longer use inline phrases like

> I'm struggling with level 4.3

Maybe we ought to remove the `^` anchor at the beginning of the match, as that seems more readable than adding a `.*`.

Also, notice that "Lol" matches. This probably isn't a good idea. Maybe we should only match "L" on its own. "Lev" probably shouldn't be matched either, but "lvl" and "level" should be.

This change will resolve some of the above:

<hr>

Also, "2019" shouldn't match. There should be a digit-boundary between each number, so that "20 19" is okay, but three-digit numbers should not work.

Here is a regular expression to consolidate the above requests:

```suggestion
            r"(le?v?e?l|l(?=\b|\d))\D*(\d{1,2})\D+(\d{1,2})\b.*$",
```

Matches
- LxCy
- lvl x y
- lvl x.y
- levl x y
- lvel x y
- level x y
- lvl x challenge y

Does not match:
- lev x y
- love x y
- level xy
- level 01 04 (note leading zeros)

We should maybe consider allowing the last one.

<hr>

Additionally, you use a capturing group to match for "level". I haven't changed it here because it would require changing some of your other code, but you never use this result so maybe we should use a non-capturing group `(?:...)` so as to only match what we need.

<hr>

I really like how this has all turned out and am open to thoughts on the above!